### ISSUE TITLE
Why Groute CC has several iteration of non-atomic hook before atomic hook?

### ISSUE THREAD

**Issue by YuxinxinChen** [groute/groute#5]
Hi,

I am reading the Groute CC code and found an interesting section and I hope to get some clarifications. The code is:  https://github.com/groute/groute/blob/develop/samples/cc/cc_common.h#L265

```c
                for (int ii = 0; ii < configuration.nonatomic_rounds; ++ii) // non atomic rounds  
                {
                    IntervalRangeMarker iter_rng(input_seg.GetSegmentSize(), "CC iteration (non-atomic)");

                    if (i == 0 && ii == 0)  problem.Work<true>(input_seg); // round 1
                    else                    problem.Work<false>(input_seg);

                    problem.CompressLocal(); 
                        // for non atomic rounds we must compress any how  
                        // compress only up to local bounds, everything else will get compressed later   
                }

                IntervalRangeMarker iter_rng(input_seg.GetSegmentSize(), "CC iteration (atomic)");


                // Work atomic
                problem.WorkAtomic(input_seg);
 
```

I also read the adaptive connected component paper: https://arxiv.org/pdf/1612.01178.pdf
In your code, there are several iterations of non-atomic hook operation before the atomic hook operation, which is not in the paper. Could you explain a little bit about why you doing this?

Thanks!

Yuxin

**Comment by michaelsutton** [comment]
We set the `nonatomic_rounds` parameter to either zero or one, depending on the average-degree of the graph. (see here: https://github.com/groute/groute/blob/develop/samples/cc/cc_config.h#L133)

If I recall correctly, the intuition was that the first Hook rounds in the original SV algorithm perform most of the work. So running a few such non-atomic rounds and then finalizing the remaining work with a single atomic round makes some sense.
However, when using segmentation (as in the arxiv paper), this makes sense only in the first segment, which is not as useful.   
We kept it only for low average-degree graphs (such as road networks) where in some cases it was beneficial.
